applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp

```diff
@@ -178,71 +179,14 @@ void UPwLysmerAbsorbingCondition<TDim, TNumNodes>::CalculateDampingMatrix(Matrix
 template< unsigned int TDim, unsigned int TNumNodes >
 void UPwLysmerAbsorbingCondition<TDim, TNumNodes>::GetValuesVector(Vector& rValues, int Step) const
 {
-
-    KRATOS_TRY
-
-        const GeometryType& r_geom = this->GetGeometry();
-
-        if (rValues.size() != CONDITION_SIZE) {
-            rValues.resize(CONDITION_SIZE, false);
-        }
-
-        if constexpr (TDim == 2) {
-            unsigned int index = 0;
-            for (unsigned int i = 0; i < TNumNodes; ++i) {
-                rValues[index++] = r_geom[i].FastGetSolutionStepValue(DISPLACEMENT_X, Step);
-                rValues[index++] = r_geom[i].FastGetSolutionStepValue(DISPLACEMENT_Y, Step);
-                rValues[index++] = 0.0;
-            }
-        }
-        else if constexpr (TDim == 3) {
-            unsigned int index = 0;
-            for (unsigned int i = 0; i < TNumNodes; ++i) {
-                rValues[index++] = r_geom[i].FastGetSolutionStepValue(DISPLACEMENT_X, Step);
-                rValues[index++] = r_geom[i].FastGetSolutionStepValue(DISPLACEMENT_Y, Step);
-                rValues[index++] = r_geom[i].FastGetSolutionStepValue(DISPLACEMENT_Z, Step);
-                rValues[index++] = 0.0;
-            }
-        }
-        else {
-            KRATOS_ERROR << "undefined dimension in GetValuesVector... illegal operation!!" << this->Id() << std::endl;
-        }
-    KRATOS_CATCH("")
+    rValues = ExtractSolutionStepValuesOfUPwDofs(this->GetDofs(), Step);
 }
 
 
 template< unsigned int TDim, unsigned int TNumNodes >
 void UPwLysmerAbsorbingCondition<TDim, TNumNodes>::GetFirstDerivativesVector(Vector& rValues, int Step) const
 {
-    KRATOS_TRY
-
-    const GeometryType& r_geom = this->GetGeometry();
-
-    if (rValues.size() != CONDITION_SIZE)
-        rValues.resize(CONDITION_SIZE, false);
-
-    if constexpr(TDim == 2) {
-        unsigned int index = 0;
-        for (unsigned int i = 0; i < TNumNodes; ++i) {
-            rValues[index++] = r_geom[i].FastGetSolutionStepValue(VELOCITY_X, Step);
-            rValues[index++] = r_geom[i].FastGetSolutionStepValue(VELOCITY_Y, Step);
-            rValues[index++] = 0.0;
-        }
-    }
-    else if constexpr (TDim == 3) {
-        unsigned int index = 0;
-        for (unsigned int i = 0; i < TNumNodes; ++i) {
-            rValues[index++] = r_geom[i].FastGetSolutionStepValue(VELOCITY_X, Step);
-            rValues[index++] = r_geom[i].FastGetSolutionStepValue(VELOCITY_Y, Step);
-            rValues[index++] = r_geom[i].FastGetSolutionStepValue(VELOCITY_Z, Step);
-            rValues[index++] = 0.0;
-        }
-    }
-    else {
-        KRATOS_ERROR << "undefined dimension in GetFirstDerivativesVector... illegal operation!!" << this->Id() << std::endl;
-    }
-
-    KRATOS_CATCH("")
+    rValues = ExtractFirstTimeDerivativesOfUPwDofs(this->GetDofs(), Step);
```

Comment:
Step here is probably 1 for start of the step ( previous value ), 0 for end of the step ( current value ). Apart from the confusing inverse setup of 0 and 1, maybe the name Step can be improved.

Reply:
I agree with you that a name such as `BufferIndex` would be much more accurate and descriptive. However, we inherit this parameter name from the Kratos core (see `condition.h`). So if we would like to change it, I feel we'd better start from there.

Also, there is an issue with the virtual function itself, since it has a default value for parameter `Step`, which opens the door to having different default values for different overrides. See for instance [this rule of the C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c140-do-not-provide-different-default-arguments-for-a-virtual-function-and-an-overrider), which warns for that. Clang-tidy also reports it:

![image](https://github.com/KratosMultiphysics/Kratos/assets/28814468/1d40baa5-6772-4280-972d-d4738e68135c)